opal/mca/common/ofi/common_ofi.c

```diff
@@ -99,87 +111,131 @@ static void opal_common_ofi_mem_release_cb(void *buf, size_t length,
 
 #endif /* HAVE_STRUCT_FI_OPS_MEM_MONITOR */
 
-int opal_common_ofi_open(void)
+int opal_common_ofi_inject_memory_monitor(void)
 {
-    int ret;
+    int ret = -FI_ENOSYS;
 
-    opal_common_ofi_init_ref_cnt++;
-    if (opal_common_ofi_initialized) {
-        return OPAL_SUCCESS;
-    }
 #ifdef HAVE_STRUCT_FI_OPS_MEM_MONITOR
+    uint32_t version;
 
-    mca_base_framework_open(&opal_memory_base_framework, 0);
+    OPAL_THREAD_LOCK(&opal_common_ofi_mutex);
+
+    if (opal_common_ofi_cache_fid != NULL) {
+        return 0;
+    }
+
+    /*
+     * While the memory import functionality was introduced in 1.13,
+     * some deadlock bugs exist in the 1.13 series.  Require version
+     * 1.14 before this code is activated.  Not activating the code
+     * should not break any functionality directly, but may lead to
+     * sub-optimal memory monitors being used in Libfabric, as Open
+     * MPI will almost certainly install a patcher first.
+     */
+    version = fi_version();
+    if (FI_VERSION_LT(version, FI_VERSION(1, 14))) {
+        ret = -FI_ENOSYS;
+        goto err;
+    }
+
+    ret = mca_base_framework_open(&opal_memory_base_framework, 0);
+    if (OPAL_SUCCESS != ret) {
+        ret = -FI_ENOSYS;
+	goto err;
+    }
     if ((OPAL_MEMORY_FREE_SUPPORT | OPAL_MEMORY_MUNMAP_SUPPORT)
```

Comment:
I think this if statement is doing the opposite of what we want to do here

Reply:
Not sure I understand.  It can't be a fatal error to us if we don't have a working memory monitor (because the MTL and BTL don't require it, although the BTL will run slower).  So we have to error and it has to be a non-fatal error.

Reply:
Ah yeah I re-read the section, ignore my comment.